File: pav_hr/pav_hr/report/cumulative_attendance_report/cumulative_attendance_report.js
// Copyright (c) 2016, Ahmed Mohammed Alkuhlani and contributors
// For license information, please see license.txt
/* eslint-disable */



frappe.query_reports["Cumulative Attendance Report"] = {
	"filters": [
		{
			fieldname: 'employee',
			label: __('Employee'),
			fieldtype: 'Link',
			options: 'Employee'
		},
		{
			"fieldname":"from",
			"label": __("From"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1
		},
		{
			"fieldname":"to",
			"label": __("To"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1
		}		
	],
	  "formatter": function (value, row, column, data, default_formatter) {
		//value = $(`<span style='font-weight:bold'>${value}</span>`);
		value = default_formatter(value, row, column, data);
		if (column.fieldname == "late_entry") 
		{
			if (data.late_entry) 
			{
				value = "<span style='background-color: #e12b2b; color:white;'><b>&nbsp;" + value + "&nbsp;</b></span>";
			}
		}
		if (column.fieldname == "early_exit")
		{
			if (data.early_exit)
			{
				value = "<span style='background-color:#e12b2b; color:white;'><b>&nbsp;" + value + "&nbsp;</b></span>";
			}
		}
		if (column.fieldname == "delay_total")
		{
			if (data.delay_total)
			{
				value = "<span style='background-color:#d00404; color:white;'><b>&nbsp;" + value + "&nbsp;</b></span>";
			}
		}
		if (column.fieldname == "early_entry") 
		{
			if (data.early_entry) 
			{
				value = "<span style='background-color:#30b530; color:white;'><b>&nbsp;" + value + "&nbsp;</b></span>";
			}
		}
		if (column.fieldname == "late_exit")
		{
			if (data.late_exit)
			{
				value = "<span style='background-color:#30b530; color:white;'><b>&nbsp;" + value + "&nbsp;</b></span>";
			}
		}
		if (column.fieldname == "early_total")
		{
			if (data.early_total)
			{
				value = "<span style='background-color:green; color:white;'><b>&nbsp;" + value + "&nbsp;</b></span>";
			}
		}
		if (column.fieldname == "working_hours")
		{
			if (data.workinghours)
			{
				value = "<span style='background-color:purple; color:white;'><b>&nbsp;" + value + "&nbsp;</b></span>";
			}
		}
		return value;
	}
};

File: pav_hr/pav_hr/report/cumulative_attendance_report/cumulative_attendance_report.py
# Copyright (c) 2021, Farouk Muharram and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import (flt,cstr)
from datetime import datetime,timedelta,time
from pav_hr.pav_hr.report.cumulative_attendance_report_with_ot.cumulative_attendance_report_with_ot import get_shift, convert_to_minutes


def delay(max, min):
	return max - min if max and min and max > min else timedelta(00,00,00)

def delayTotal(max, min):
	if max and min:
		return max + min
	elif max and not min:
		return max
	elif min and not max:
		return min 
	else:
		timedelta(00,00,00)

def execute(filters=None):
	frappe.msgprint("123")
	if not filters: filters = {}
	formatted_data = []
	columns = get_columns()
	data = get_data(filters)
	shift = get_shift()
	for emp in data:
		basic_total = None
		over_total = None
		startLate = None
		endLate = None
		tLate = None
		startEr = None
		endEr = None
		tEr = None
		for d in data[emp]:
			if d[7] == "NOT NULL":
				temp= delay(d[4], d[3])
				basic_total = delayTotal(basic_total, temp)

				d3 = d[3]
				d6 = d[6]
				for sh in shift:
					entry_period = convert_to_minutes(sh[1])
					exit_period = convert_to_minutes(sh[2])
					if sh[0] == d[2]:
						if d[3] < delayTotal(d[5], entry_period):
							startLate = startLate
						else:
							temp= delay(d3, d[5])
							startLate = delayTotal(startLate, temp)
						
						if d[4] >= delayTotal(d[6], exit_period):
							endLate = endLate
						else:
							temp = delay(d6, d[4])
							endLate = delayTotal(endLate, temp)

				temp = delay(d[5], d[3])
				startEr = delayTotal(startEr, temp)
				temp = delay(d[4], d[6])
				endEr = delayTotal(endEr, temp)					
			elif d[7] == "NULL":
				temp= delay(d[4], d[3])
				over_total = delayTotal(over_total, temp)

		tLate = delayTotal(startLate, endLate)
		tEr = delayTotal(startEr, endEr)
		tHours = delayTotal(basic_total, over_total)
		formatted_data.append({
		"name": emp,
		"employee": d[0],
		"late_entry":   to_hours(startLate) if startLate else startLate,
		"early_exit":  to_hours(endLate) if endLate else endLate,
		"late_total" :  to_hours(tLate) if tLate else tLate,
		"early_entry" :  to_hours(startEr) if startEr else startEr,
		"late_exit" : to_hours(endEr) if endEr else endEr,
		"early_total" :  to_hours(tEr) if tEr else tEr,
		"working_hours": to_hours(tHours) if tHours else tHours,
			})
				
	formatted_data.extend([{}])
	return columns, formatted_data

def to_hours(duration):
	if duration:
		totsec = duration.total_seconds()
		h = totsec//3600
		m = (totsec%3600) // 60
		sec =(totsec%3600)%60
		return "%d:%d:%d" %(h,m,sec)
	else:
		return timedelta(00,00,00)

def get_columns():
	return [
		{
			"fieldname": "name",
			"label": _("Employee "),
			"fieldtype": "Link",
			"options": "Employee",
			"width": 120
		},
		{
			"fieldname": "employee",
			"label": _("Employee Name"),
			"fieldtype": "Data",
			"width": 170
		},
		{
			"fieldname": "late_entry",
			"label": _("Late Entry"),
			"fieldtype": "Data",
			"width": 80
		},
		{
			"fieldname": "early_exit",
			"label": _("Early Exit"),
			"fieldtype": "Data",
			"width": 80
		},
		      {
			"fieldname": "late_total",
			"label": _("Late Total"),
			"fieldtype": "Data",
			"width": 85
		},
		 {
			"fieldname": "early_entry",
			"label": _("Early Entry"),
			"fieldtype": "Data",
			"width": 80
		 },
		 		{
			"fieldname": "late_exit",
			"label": _("Late Exit"),
			"fieldtype": "Data",
			"width": 80
		},
		{
			"fieldname": "early_total",
			"label": _("Early Total"),
			"fieldtype": "Data",
			"width": 85
		},
		 {
			"fieldname": "working_hours",
			"label": _("Working Hours"),
			"fieldtype": "Data",
			"width": 120
		},
	]
	
def get_conditions(filters):
	condition = " where date(time)>= '%s' and date(time)<= '%s' " % (filters.get("from"), filters.get("to"))
	if filters.get("employee"):
		condition += " AND employee ='%s' " % (filters.get('employee'))
	return condition


def get_data(filters):
	list_ = frappe.db.sql('''SELECT employee, employee_name, date(time) as ckin_date, shift, TIME(min(time)) as in_time,
		TIME(max(time)) as out_time, TIME(MAX(shift_start)) as shift_start, TIME(MAX(shift_end)) as shift_end, 
		IF(ISNULL(shift), 'NULL', 'NOT NULL') as shift_nullable
		FROM `tabEmployee Checkin` {0} GROUP BY employee, date(time), shift_nullable'''.format(get_conditions(filters)), as_list=1)

	result = {}
	for d in list_:
		result.setdefault(d[0], [])
		result[d[0]].append([d[1], d[2], d[3], d[4], d[5], d[6], d[7], d[8]])

	return result

